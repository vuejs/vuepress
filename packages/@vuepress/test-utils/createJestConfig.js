const defaultJestConfig = require('./jest/jest.config')

module.exports = function createJestConfig (override) {
  return Object.assign({}, defaultJestConfig, override)
}
