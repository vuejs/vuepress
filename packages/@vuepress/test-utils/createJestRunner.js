const execa = require('execa')
const rawArgs = process.argv.slice(2)

module.exports = function createJestRunner (jestArgs) {
  return async function () {
    const args = [...jestArgs, ...rawArgs]
    console.log(`running jest with args: ${args.join(' ')}`)
    await execa('jest', args, {
      stdio: 'inherit'
    })
  }
}
