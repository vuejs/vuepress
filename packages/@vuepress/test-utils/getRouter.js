const Router = require('vue-router')

module.exports = function () {
  return new Router()
}
