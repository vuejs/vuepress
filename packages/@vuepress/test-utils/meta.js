const { fs } = require('@vuepress/shared-utils')
const path = require('path')

const docsBaseDir = path.resolve(__dirname, 'fixtures')
const docsModeNames = fs.readdirSync(docsBaseDir)
const docsModes = docsModeNames.map(name => {
  const docsPath = path.resolve(docsBaseDir, name)
  const docsTempPath = path.resolve(docsPath, '.vuepress/.temp')
  return { name, docsPath, docsTempPath }
})

exports.docsModes = docsModes
