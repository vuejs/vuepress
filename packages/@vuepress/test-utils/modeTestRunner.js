const { createLocalVue } = require('@vue/test-utils')
const dataMixin = require('@vuepress/core/lib/app/dataMixin').default
const I18n = require('@vuepress/core/lib/prepare/I18n')
const Router = require('vue-router')
const mockComponent = require('./mockComponent')
const { docsModes } = require('./meta')

function getLocalVueByMode (mode) {
  const localVue = createLocalVue()
  localVue.use(Router)

  // register global component
  localVue.component('OutboundLink', mockComponent('outbound-link'))

  const { siteData } = require(`${mode.docsTempPath}/internal/siteData.js`)
  localVue.component(siteData.pages[0].key, mockComponent('page-component'))
  localVue.mixin(dataMixin(I18n, siteData))
  return localVue
}

/**
 * Used to test components in VuePress at different mode.
 * Since VuePress has two main modes, i18n or normal mode, we need to make
 * sure that the core Vue components in VuePress work well in both modes.
 * @param description
 * @param testFn
 */
module.exports = function modeTestRunner (description, testFn) {
  docsModes.forEach(mode => {
    describe(description, () => {
      testFn(mode.name, getLocalVueByMode(mode))
    })
  })
}
