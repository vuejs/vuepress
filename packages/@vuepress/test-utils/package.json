{
  "name": "@vuepress/test-utils",
  "version": "1.0.0",
  "description": "test-utils for vuepress",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuejs/vuepress.git"
  },
  "keywords": [
    "documentation",
    "vue",
    "vuepress",
    "generator"
  ],
  "dependencies": {
    "@vuepress/core": "^1.0.0",
    "@vuepress/shared-utils": "^1.0.0",
    "@vue/test-utils": "^1.0.0-beta.16",
    "@babel/core": "^7.0.0-beta.47",
    "@babel/preset-env": "^7.0.0-beta.47",
    "babel-core": "^7.0.0-0",
    "babel-jest": "^23.4.0",
    "jest": "^23.4.0",
    "jest-serializer-vue": "^1.0.0",
    "vue-jest": "^2.6.0",
    "execa": "^0.10.0"
  },
  "author": "Evan You",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vuejs/vuepress/issues"
  },
  "homepage": "https://github.com/vuejs/vuepress/packages/@vuepress/test-utils#readme"
}
