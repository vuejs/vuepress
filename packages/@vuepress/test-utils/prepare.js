const { fs, logger } = require('@vuepress/shared-utils')
const prepare = require('@vuepress/core/lib/prepare')
const { docsModes } = require('./meta')

async function prepareForTest () {
  await Promise.all(docsModes.map(async ({ docsPath, docsTempPath }) => {
    await fs.ensureDir(docsTempPath)
    await prepare(docsPath, { theme: '@vuepress/default', temp: docsTempPath })
  }))
}

module.exports = function globalSetup () {
  logger.wait('Start preparing for testing ...')
  return prepareForTest().then(() => {
    logger.wait('Finished preparing for testing ...')
  })
}
