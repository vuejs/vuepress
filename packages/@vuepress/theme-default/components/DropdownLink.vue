<template>
  <div
    class="dropdown-wrapper"
    :class="{ open }"
  >
    <button
      class="dropdown-title"
      type="button"
      :aria-label="dropdownAriaLabel"
      @click="handleDropdown"
    >
      <span class="title">{{ item.text }}</span>
      <span
        class="arrow down"
      />
    </button>
    <button
      class="mobile-dropdown-title"
      type="button"
      :aria-label="dropdownAriaLabel"
      @click="setOpen(!open)"
    >
      <span class="title">{{ item.text }}</span>
      <span
        class="arrow"
        :class="open ? 'down' : 'right'"
      />
    </button>

    <DropdownTransition>
      <ul
        v-show="open"
        class="nav-dropdown"
      >
        <li
          v-for="(subItem, index) in item.items"
          :key="subItem.link || index"
          class="dropdown-item"
        >
          <h4 v-if="subItem.type === 'links'">
            {{ subItem.text }}
          </h4>

          <ul
            v-if="subItem.type === 'links'"
            class="dropdown-subitem-wrapper"
          >
            <li
              v-for="childSubItem in subItem.items"
              :key="childSubItem.link"
              class="dropdown-subitem"
            >
              <NavLink
                :item="childSubItem"
                @focusout="
                  isLastItemOfArray(childSubItem, subItem.items) &&
                    isLastItemOfArray(subItem, item.items) &&
                    setOpen(false)
                "
              />
            </li>
          </ul>

          <NavLink
            v-else
            :item="subItem"
            @focusout="isLastItemOfArray(subItem, item.items) && setOpen(false)"
          />
        </li>
      </ul>
    </DropdownTransition>
  </div>
</template>

<script>
import NavLink from '@theme/components/NavLink.vue'
import DropdownTransition from '@theme/components/DropdownTransition.vue'
import last from 'lodash/last'

export default {
  name: 'DropdownLink',

  components: {
    NavLink,
    DropdownTransition
  },

  props: {
    item: {
      required: true
    }
  },

  data () {
    return {
      open: false
    }
  },

  computed: {
    dropdownAriaLabel () {
      return this.item.ariaLabel || this.item.text
    }
  },

  watch: {
    $route () {
      this.open = false
    }
  },

  methods: {
    setOpen (value) {
      this.open = value
    },

    isLastItemOfArray (item, array) {
      return last(array) === item
    },

    /**
     * Open the dropdown when user tab and click from keyboard.
     *
     * Use event.detail to detect tab and click from keyboard. Ref: https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/detail
     * The Tab + Click is UIEvent > KeyboardEvent, so the detail is 0.
     */
    handleDropdown () {
      const isTriggerByTab = event.detail === 0
      if (isTriggerByTab) this.setOpen(!this.open)
    }
  }
}
</script>

<style lang="stylus">
.dropdown-wrapper
  cursor pointer
  .dropdown-title
    display block
    font-size 0.9rem
    font-family inherit
    cursor inherit
    padding inherit
    line-height 1.4rem
    background transparent
    border none
    font-weight 500
    color $textColor
    &:hover
      border-color transparent
    .arrow
      vertical-align middle
      margin-top -1px
      margin-left 0.4rem
  .mobile-dropdown-title
    @extends .dropdown-title
    display none
    font-weight 600
    font-size inherit
      &:hover
        color $accentColor
  .nav-dropdown
    .dropdown-item
      color inherit
      line-height 1.7rem
      h4
        margin 0.45rem 0 0
        border-top 1px solid #eee
        padding 1rem 1.5rem 0.45rem 1.25rem
      .dropdown-subitem-wrapper
        padding 0
        list-style none
        .dropdown-subitem
          font-size 0.9em
      a
        display block
        line-height 1.7rem
        position relative
        border-bottom none
        font-weight 400
        margin-bottom 0
        padding 0 1.5rem 0 1.25rem
        &:hover
          color $accentColor
        &.router-link-active
          color $accentColor
          &::after
            content ""
            width 0
            height 0
            border-left 5px solid $accentColor
            border-top 3px solid transparent
            border-bottom 3px solid transparent
            position absolute
            top calc(50% - 2px)
            left 9px
      &:first-child h4
        margin-top 0
        padding-top 0
        border-top 0

@media (max-width: $MQMobile)
  .dropdown-wrapper
    &.open .dropdown-title
      margin-bottom 0.5rem
    .dropdown-title
      display: none
    .mobile-dropdown-title
      display: block
    .nav-dropdown
      transition height .1s ease-out
      overflow hidden
      .dropdown-item
        h4
          border-top 0
          margin-top 0
          padding-top 0
        h4, & > a
          font-size 15px
          line-height 2rem
        .dropdown-subitem
          font-size 14px
          padding-left 1rem

@media (min-width: $MQMobile)
  .dropdown-wrapper
    height 1.8rem
    &:hover .nav-dropdown,
    &.open .nav-dropdown
      // override the inline style.
      display block !important
    .nav-dropdown
      display none
      // Avoid height shaked by clicking
      height auto !important
      box-sizing border-box;
      max-height calc(100vh - 2.7rem)
      overflow-y auto
      position absolute
      top 100%
      right 0
      background-color #fff
      padding 0.6rem 0
      border 1px solid #ddd
      border-bottom-color #ccc
      text-align left
      border-radius 0.25rem
      white-space nowrap
      margin 0
</style>
