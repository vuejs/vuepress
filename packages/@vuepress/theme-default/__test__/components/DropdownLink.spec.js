import { mount, RouterLinkStub } from '@vue/test-utils'
import DropdownLink from '../../components/DropdownLink.vue'
import modeTestRunner from '@vuepress/test-utils/modeTestRunner'

function test (mode, localVue) {
  it(`$${mode} - renders dropdown link.`, () => {
    const item = {
      text: 'VuePress',
      items: [
        {
          text: 'Guide',
          link: '/guide/'
        },
        {
          text: 'Config Reference',
          link: '/config/'
        }
      ]
    }
    const wrapper = mount(DropdownLink, {
      localVue,
      stubs: {
        'router-link': RouterLinkStub
      },
      propsData: { item }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
}

modeTestRunner('DropdownLink', test)
