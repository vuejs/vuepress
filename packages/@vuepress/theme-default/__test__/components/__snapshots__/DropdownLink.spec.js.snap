// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownLink $docs-i18n - renders dropdown link. 1`] = `
<div class="dropdown-wrapper">
  <a class="dropdown-title"><span class="title">VuePress</span> <span class="arrow right"></span></a>
  <ul class="nav-dropdown" style="display: none;" name="dropdown">
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Guide</a>
    </li>
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Config Reference</a>
    </li>
  </ul>
</div>
`;

exports[`DropdownLink $docs-local-theme - renders dropdown link. 1`] = `
<div class="dropdown-wrapper">
  <a class="dropdown-title"><span class="title">VuePress</span> <span class="arrow right"></span></a>
  <ul class="nav-dropdown" style="display: none;" name="dropdown">
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Guide</a>
    </li>
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Config Reference</a>
    </li>
  </ul>
</div>
`;

exports[`DropdownLink $docs-simple - renders dropdown link. 1`] = `
<div class="dropdown-wrapper">
  <a class="dropdown-title"><span class="title">VuePress</span> <span class="arrow right"></span></a>
  <ul class="nav-dropdown" style="display: none;" name="dropdown">
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Guide</a>
    </li>
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Config Reference</a>
    </li>
  </ul>
</div>
`;

exports[`DropdownLink $docs-simple-config - renders dropdown link. 1`] = `
<div class="dropdown-wrapper">
  <a class="dropdown-title"><span class="title">VuePress</span> <span class="arrow right"></span></a>
  <ul class="nav-dropdown" style="display: none;" name="dropdown">
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Guide</a>
    </li>
    <li class="dropdown-item">
      <!---->
      <a class="nav-link">Config Reference</a>
    </li>
  </ul>
</div>
`;
